File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        res = 0
        
        mp = {}
        i = 0
        
        for j in range(n):
            char = s[j]
            if char in mp:
                i = max(mp[char], i)
            
            res = max(res, j - i + 1)
            mp[char] = j + 1
        
        return res

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let chars = new Map()
    let max = 0
    let start = 0
    for (let i = 0; i < s.length; i++) {
        const ch = s[i]
        if (chars.get(ch) >= start) start = chars.get(ch) + 1
        chars.set(ch, i)
        max = Math.max(max, i - start + 1)
    }
    return max
};